keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt

```diff
@@ -50,158 +39,176 @@ import java.util.function.BiFunction
+          .toSet()
+      }
+        .getOrElse { ex ->
+          throw CacheLoadingException("Error loading cache", ex)
+        }
+    }
 
-  private val namedImagesByAccountAndImageName = buildAsyncCache<String, List<NamedImage>>(
-    cacheName = "namedImages",
-    defaultExpireAfterWrite = Duration.ofMinutes(5)
-  )
+  private val credentials: AsyncLoadingCache<String, Credential> = cacheFactory
+    .asyncLoadingCache(
+      cacheName = "credentials"
+    ) { name ->
+      runCatching {
+        cloudDriver.getCredential(name, DEFAULT_SERVICE_ACCOUNT)
+      }
+        .handleNotFound()
+        ?: notFound("Credentials with name $name not found")
+    }
 
-  override fun credentialBy(name: String): Credential =
-    credentials.getOrNotFound(name, "Credentials with name $name not found") {
-      cloudDriver.getCredential(name, DEFAULT_SERVICE_ACCOUNT)
+  private val subnetsById: AsyncLoadingCache<String, Subnet> = cacheFactory
+    .asyncLoadingCache(cacheName = "subnetsById") { subnetId ->
+      runCatching {
+        cloudDriver
+          .listSubnets("aws", DEFAULT_SERVICE_ACCOUNT)
+          .find { it.id == subnetId }
+      }
+        .handleNotFound()
+        ?: notFound("Subnet with id $subnetId not found")
     }
 
-  /**
-   * Convert a suspending function to a [BiFunction] that [AsyncCache.get] expects as its second argument
-   *
-   * Uses a coroutine dispatcher based on the cache's [Executor].
-   */
-  private fun <K, V> asyncify(block: suspend CoroutineScope.(K) -> V?): BiFunction<K, Executor, CompletableFuture<V?>> =
-    BiFunction { key, executor ->
-      CoroutineScope(executor.asCoroutineDispatcher())
-        .future(block = { block(key) })
+  private val subnetsByPurpose: AsyncLoadingCache<Triple<String, String, String>, Subnet> = cacheFactory
+    .asyncLoadingCache(cacheName = "subnetsByPurpose") { (account, region, purpose) ->
+      runCatching {
+        cloudDriver
+          .listSubnets("aws", DEFAULT_SERVICE_ACCOUNT)
+          .find { it.account == account && it.region == region && it.purpose == purpose }
+      }
+        .handleNotFound()
+        ?: notFound("Subnet with purpose \"$purpose\" not found in $account:$region")
+    }
+
+  override fun credentialBy(name: String): Credential =
+    runBlocking {
+      credentials.get(name).await()
     }
 
   override fun securityGroupById(account: String, region: String, id: String): SecurityGroupSummary =
-    securityGroupSummariesByIdOrName.getOrNotFound(
-      "$account:$region:$id",
-      "Security group with id $id not found in the $account account and $region region"
-    ) {
-      val credential = credentialBy(account)
-      cloudDriver.getSecurityGroupSummaryById(account, credential.type, region, id, DEFAULT_SERVICE_ACCOUNT)
-    }.also {
-      securityGroupSummariesByIdOrName.synchronous().put("$account:$region:${it.name}", it)
+    runBlocking {
+      securityGroupsById.get(Triple(account, region, id)).await()
     }
```

Comment:
The logic from all these methods is moved into the cache loader so here it's just a simple cache get.